### ISSUE TITLE
Some styles have bad contrasts

### ISSUE THREAD

**Issue by not-my-profile** [pygments/pygments#1718]
After encountering a bad color contrast in the syntax highlighting of the [official PIP documentation](https://pip.pypa.io/en/stable/reference/pip_show/#examples),

![image](https://user-images.githubusercontent.com/73739153/108076429-b545e380-706b-11eb-8c10-10c0a005711a.png)

I wrote [a small Python script](https://git.push-f.com/tools/tree/pygments_wcag_check.py) to check how Pygments styles fare when judged according to the [WCAG contrast standard](https://www.w3.org/WAI/WCAG21/Understanding/contrast-minimum.html).

The summary of my script is:
```
found 840 contrasts that fail to meet the WCAG AA standard
According to WCAG:
     AA contrast is >= 4.5
    AAA contrast is >= 7.0
    
=== Styles ranked by contrast ===
 avg   min  name
 3.4   2.5  solarized-light
 4.2   1.4  stata-dark
 4.3   2.8  solarized-dark
 5.5   1.3  paraiso-light
 6.2   1.3  paraiso-dark
 6.3   1.0  rrt
 7.0   1.0  inkpot
 7.3   1.7  friendly
 7.5   1.6  perldoc
 7.5   2.1  native
 7.6   1.8  emacs
 7.8   1.3  vim
 7.9   1.4  monokai
 8.0   1.4  manni
 8.7   3.1  lovelace
 8.7   1.8  default
 8.9   2.6  arduino
 9.2   1.9  autumn
 9.3   1.9  pastie
 9.6   1.0  rainbow_dash
 9.9   1.9  trac
10.1   1.4  murphy
10.4   1.0  tango
10.7   1.9  fruity
10.7   1.9  colorful
11.1   1.9  sas
13.0   2.5  abap
13.2   1.9  stata
13.2   1.9  stata-light
13.8   1.9  borland
13.9   5.1  xcode
14.4   3.6  igor
14.7   3.5  algol
14.7   3.5  algol_nu
15.7   3.6  vs
21.0  21.0  bw
```

My script also reports which rules don't meet the AA standard per style sorted by contrast, e.g. for the default style:
```
=== default ===
1.81 Token.Text.Whitespace
2.63 Token.Name.Label
2.68 Token.Name.Entity
3.28 Token.Generic.Inserted
3.34 Token.Generic.Output
3.34 Token.Comment.Preproc
3.36 Token.Name.Attribute
3.68 Token.Literal.String.Interpol
3.68 Token.Literal.String.Regex
3.76 Token.Generic.Error
3.93 Token.Literal.String.Escape
4.28 Token.Comment
4.28 Token.Comment.Hashbang
4.28 Token.Comment.Multiline
4.28 Token.Comment.PreprocFile
4.28 Token.Comment.Single
4.28 Token.Comment.Special
4.34 Token.Name.Exception
```

I am not saying that every Pygments style needs to have an AA contrast for every token type, but I think at least the default style should, since it's probably the most used one. If you want to I can send a PR to add my script into the `scripts` directory.

To improve the accessibility long term, I think it would make sense to incorporate the script into your CI pipeline with a given minimum contrast, making the pipeline fail if any style fails to meet this minimum. As the worst contrasts are improved this minimum contrast could then be gradually raised.

**Comment by Anteru** [comment]
Thanks! This is actually quite important and I don't think we've been paying nearly enough attention to this. The real question is how to make sure this is checked in CI and we don't degrade. The way I'd like to see this work is to be embedded with our current test infrastructure, and have the current minimum value stored for each theme as the baseline -- this way we can prevent regressions. If a theme is unknown, then the minimum should be set to the WCAG recommendation -- a theme author could still bypass it, but that would be a manual process. Do you think that's something you can work on? I believe it should be "as easy" as writing a new test which iterates over each style, applies the function from your script, and checks against the current value for the theme (and defaulting to 4.5 if it's an unknown theme.)

I don't think we can hope to fix all themes simply because some of them are designed to be low-contrast, but we could (should!) advertise accessible themes in the hope more people will use them, and have a minimum bar for new themes.

**Comment by not-my-profile** [comment]
Thanks for your answer. Yeah I should be able to find some time for this in the next months but I can't promise. I assume to check against the current value for a theme, the current values would need to be tracked in the repository (in e.g. a JSON file)?

It's worth noting that the complete lack of contrast (min=1) reported for some styles is apparently caused by missing definitions (e.g. black text on black background). So these should really be fixed for all styles.

**Comment by Anteru** [comment]
Yes, either in a JSON file or simply directly in the test file (as a dictionary of style name & min contrast.) You're also absolutely right about the missing definitions that should be fixed, if we can identify those automatically that would help a lot. My goal would be to have a long-term strategy in place to avoid style regressions, and also make it easier for new styles developers to "sanity check" their new theme in terms of accessibility. I don't think we can fix all styles (some are deliberately low contrast) but we can definitely highlight accessible styles.

**Comment by yu-andy** [comment]
Hi @not-my-profile the link to your Python script results in a 502 Bad Gateway Error, could you host it somewhere else? I’d like to try test my theme out :)

**Comment by not-my-profile** [comment]
Hi, thanks for the notice! I fixed the 502 :)

**Comment by not-my-profile** [comment]
@Anteru I finally found the time & energy to implement this, only took me 8 months^^. Let me know if this is what you had in mind :)

I think the next two steps would be:

* generating a table for the website that shows styles ranked by their minimum contrast
* making the default style meet the AA standard

**Comment by jonatanschroeder** [comment]
Just adding a note to this issue (since I assume it's related): I couldn't find any dark background style that works properly with the "Text" lexer, at least using the HTML formatter. All of them give me really low contrast, some no contrast at all.
